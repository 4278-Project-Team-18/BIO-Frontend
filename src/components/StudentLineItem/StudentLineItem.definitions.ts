import type { Student } from "../../interfaces/user.interface";

export interface StudentLineItemProps {
  student: Student;
  index: number;
  maxIndex: number;
}
