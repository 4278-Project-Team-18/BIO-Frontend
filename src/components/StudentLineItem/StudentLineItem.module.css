@media (min-width: 600px) {
  .line-item-container {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    background-color: white;
    border-bottom: 1px solid #f2f2f2;
    padding: 5px;
  }

  .line-item-container-first {
    border-radius: 20px 20px 0px 0px;
  }

  .line-item-container-last {
    border-radius: 0px 0px 20px 20px;
  }

  .line-item-container-left {
    display: flex;
    justify-content: left;
    align-items: center;
    padding: 5px;
  }

  .line-item-name {
    font-weight: 400;
    font-size: 16px;
    color: #565656;
  }

  .line-item-container-right {
    display: flex;
    justify-content: left;
    align-items: center;
    padding: 5px;
  }

  .line-item-upload-button {
    display: flex;
    justify-content: left;
    align-items: center;
    padding-top: 2px;
  }

  .line-item-upload-title {
    font-weight: 300;
    font-size: 14px;
    color: #434343;
    font-family: "Poppins", sans-serif;
    font-weight: 300;
  }

  .line-item-upload-icon {
    font-size: 16px;
    padding-left: 10px;
    padding-bottom: 2px;
    color: #434343;
  }

  .line-item-checkbox {
    font-size: 24px;
    padding-left: 10px;
    padding-right: 30px;
  }

  .line-item-checkbox-checked {
    color: #75d887;
  }

  .line-item-checkbox-unchecked {
    color: #646464;
  }
}

@media (max-width: 599px) {
  .line-item-container {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    background-color: white;
    border-bottom: 1px solid #f2f2f2;
    padding: 5px;
  }

  .line-item-container-first {
    border-radius: 20px 20px 0px 0px;
  }

  .line-item-container-last {
    border-radius: 0px 0px 20px 20px;
  }

  .line-item-container-left {
    display: flex;
    justify-content: left;
    align-items: center;
    padding: 5px;
  }

  .line-item-name {
    font-weight: 400;
    font-size: 14px;
    color: #565656;
  }

  .line-item-container-right {
    display: flex;
    justify-content: left;
    align-items: center;
    padding: 5px;
  }

  .line-item-upload-button {
    display: flex;
    justify-content: left;
    align-items: center;
    padding-top: 2px;
  }

  .line-item-upload-title {
    display: none;
  }

  .line-item-upload-icon {
    font-size: 16px;
    padding-left: 10px;
    padding-bottom: 2px;
    color: #434343;
  }

  .line-item-checkbox {
    font-size: 18px;
    padding-left: 10px;
    padding-right: 20px;
  }

  .line-item-checkbox-checked {
    color: #75d887;
  }

  .line-item-checkbox-unchecked {
    color: #646464;
  }
}
