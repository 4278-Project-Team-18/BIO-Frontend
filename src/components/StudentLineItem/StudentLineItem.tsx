import styles from "./StudentLineItem.module.css";
import LineItemLabel from "../FileUploadedLabel/LineItemLabel";
import { LineItemLabelVariant } from "../FileUploadedLabel/LineItemLabel.definitions";
import UploadStudentLetterModal from "../../modals/UploadStudentLetter/UploadStudentLetterModal";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCloudArrowUp,
  faPencil,
  faTrash,
} from "@fortawesome/free-solid-svg-icons";
import {
  faSquare,
  faSquareCheck,
  faCircleCheck,
} from "@fortawesome/free-regular-svg-icons";
import { MoonLoader } from "react-spinners";
import { useState } from "react";
import type { StudentLineItemProps } from "./StudentLineItem.definitions";

const StudentLineItem = ({
  student,
  openEditModal,
  removeStudentFromClass,
  removeStudentLoading = false,
}: StudentLineItemProps) => {

  const [uploadLetterModalOpen, setUploadLetterModalOpen] = useState<boolean>(false);

  const handleCloseUploadLetterModal = () => {
    setUploadLetterModalOpen(false);
  };

  const handleOpenUploadLetterModal = () => {
    setUploadLetterModalOpen(true);
  };

   return (<div className={styles["line-item-container"]}>
   <div className={styles["line-item-container-left"]}>
     <div className={styles["line-item-checkbox"]}>
       {student.studentLetterLink ? (
         <FontAwesomeIcon
           icon={faSquareCheck}
           className={styles["line-item-checkbox-checked"]}
         />
       ) : (
         <FontAwesomeIcon
           icon={faSquare}
           className={styles["line-item-checkbox-unchecked"]}
         />
       )}
     </div>
     <div className={styles["line-item-student-info"]}>
       <div
         className={styles["line-item-name"]}
       >{`${student.firstName} ${student.lastInitial}`}</div>
       <div
         className={styles["line-item-reading-level"]}
       >{`Reading level: ${student.readingLevel}`}</div>
     </div>
   </div>
   <div className={styles["line-item-container-right"]}>
     {student.studentLetterLink && (
       <LineItemLabel
         variant={LineItemLabelVariant.GREEN}
         label={" Student Letter"}
         icon={faCircleCheck}
       />
     )}
     <button className={styles["line-item-upload-button"]}
     onClick={handleOpenUploadLetterModal}>
       <div className={styles["line-item-upload-title"]}>
         {student.studentLetterLink ? "Reupload" : "Upload"}
       </div>
       <FontAwesomeIcon
         icon={faCloudArrowUp}
         className={styles["line-item-upload-icon"]}
       />
       {uploadLetterModalOpen && <UploadStudentLetterModal closeModal={handleCloseUploadLetterModal} student={student}/>}
     </button>
     <div className={styles["line-item-actions"]}>
       <button
         className={styles["line-item-edit-button"]}
         onClick={() => openEditModal(student._id)}
       >
         <FontAwesomeIcon icon={faPencil} />
       </button>
       {removeStudentLoading ? (
         <div className={styles["line-item-remove-button-spinner"]}>
           <MoonLoader color="#434343" size={16} />
         </div>
       ) : (
         <button
           className={styles["line-item-remove-button"]}
           onClick={() => removeStudentFromClass(student._id)}
         >
           <FontAwesomeIcon icon={faTrash} />
         </button>
       )}
     </div>
   </div>
 </div>
);


}

export default StudentLineItem;
