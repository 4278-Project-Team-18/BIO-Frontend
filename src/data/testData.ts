import {
  type Admin,
  type Student,
  type Class,
  type Teacher,
  type Volunteer,
  Role,
  ApprovalStatus,
} from "../interfaces/user.interface";
import { faker } from "@faker-js/faker";

export const createTestAdmin = () =>
  ({
    _id: faker.string.alphanumeric(26),
    firstName: faker.person.firstName(),
    lastName: faker.person.lastName(),
    email: faker.internet.email(),
    password: faker.internet.password(),
    role: Role.ADMIN,
    authToken: faker.string.uuid(),
    refreshToken: faker.string.uuid(),
  }) as Admin;

/**
 * @function createTestStudent
 * @param isTest - optional parameter to determine if the image link should always be generated or not (for snapshots)
 * @returns a student object with random data
 */
export const createTestStudent = () =>
  ({
    _id: faker.string.alphanumeric(26),
    firstName: faker.person.firstName(),
    lastInitial: faker.person.lastName().charAt(0),
    readingLevel: faker.number.bigInt({ min: 100, max: 1500 }).toString(),
    studentLetterLink: randomImageLink(),
    volunteerResponseLetterLink: randomImageLink(),
  }) as Student;

/**
 * @function createTestClass
 * @returns a class object with random data
 */
export const createTestClass = () =>
  ({
    _id: faker.string.alphanumeric(26),
    name: faker.word.adjective() + " " + faker.word.noun() + " Class",
    teacherId: faker.string.alphanumeric(26),
  }) as Class;

/**
 * @function createTestTeacher
 * @returns a teacher object with random data
 */
export const createTestTeacher = () =>
  ({
    _id: faker.string.alphanumeric(26),
    firstName: faker.person.firstName(),
    lastName: faker.person.lastName(),
    email: faker.internet.email(),
    password: faker.internet.password(),
    role: Role.TEACHER,
    authToken: faker.string.uuid(),
    refreshToken: faker.string.uuid(),
    approvalStatus: randomApprovalStatus(),
  }) as Teacher;

/**
 * @function createTestVolunteer
 * @returns a volunteer object with random data
 */
export const createTestVolunteer = () =>
  ({
    _id: faker.string.alphanumeric(26),
    firstName: faker.person.firstName(),
    lastName: faker.person.lastName(),
    email: faker.internet.email(),
    password: faker.internet.password(),
    role: Role.VOLUNTEER,
    authToken: faker.string.uuid(),
    refreshToken: faker.string.uuid(),
    approvalStatus: randomApprovalStatus(),
  }) as Volunteer;

/**
 * @function randomApprovalStatus
 * @returns a random approval status
 */
export const randomApprovalStatus = () => {
  const approvalStatuses = [
    ApprovalStatus.APPROVED,
    ApprovalStatus.PENDING,
    ApprovalStatus.REJECTED,
  ];
  return approvalStatuses[Math.floor(Math.random() * approvalStatuses.length)];
};

export const randomImageLink = () => {
  if (faker.number.int(1) > 5) {
    return faker.internet.url();
  } else {
    return undefined;
  }
};
