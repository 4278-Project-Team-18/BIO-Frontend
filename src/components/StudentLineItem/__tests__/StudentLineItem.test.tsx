import StudentLineItem from "../StudentLineItem";
import { createTestStudent } from "../../../data/testData";
import { render } from "@testing-library/react";
import { faker } from "@faker-js/faker";

describe("StudentLineItem Tests", () => {
  test("StudentLineItem should render correctly", () => {
    const testStudent = createTestStudent();

    const component = render(<StudentLineItem student={testStudent} />);

    // Check if the component is rendered
    expect(component).toBeTruthy();

    // Check if the label is correct
    expect(
      component.getByText(
        `${testStudent.firstName} ${testStudent.lastInitial}`,
      ),
    ).toBeTruthy();

    // Check if the upload button is correct
    if (
      !testStudent.volunteerResponseLetterLink ||
      testStudent.volunteerResponseLetterLink === ""
    ) {
      expect(component.getByText("Upload Letter")).toBeTruthy();
    } else {
      expect(component.getByText("Reupload")).toBeTruthy();
    }
  });

  test("StudentLineItem snapshot should render correctly", () => {
    faker.seed(1);
    const testStudent = createTestStudent();

    const component = render(<StudentLineItem student={testStudent} />);

    expect(component).toMatchSnapshot();
  });
});
