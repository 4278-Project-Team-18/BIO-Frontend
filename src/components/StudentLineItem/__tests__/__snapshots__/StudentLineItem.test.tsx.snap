// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentLineItem Tests StudentLineItem snapshot should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="line-item-container"
      >
        <div
          class="line-item-container-left"
        >
          <div
            class="line-item-checkbox"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-square line-item-checkbox-unchecked"
              data-icon="square"
              data-prefix="far"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M384 80c8.8 0 16 7.2 16 16V416c0 8.8-7.2 16-16 16H64c-8.8 0-16-7.2-16-16V96c0-8.8 7.2-16 16-16H384zM64 32C28.7 32 0 60.7 0 96V416c0 35.3 28.7 64 64 64H384c35.3 0 64-28.7 64-64V96c0-35.3-28.7-64-64-64H64z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="line-item-name"
          >
            Shad B
          </div>
        </div>
        <div
          class="line-item-container-right"
        >
          <button
            class="line-item-upload-button"
          >
            <div
              class="line-item-upload-title"
            >
              Upload Letter
            </div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-cloud-arrow-up line-item-upload-icon"
              data-icon="cloud-arrow-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M144 480C64.5 480 0 415.5 0 336c0-62.8 40.2-116.2 96.2-135.9c-.1-2.7-.2-5.4-.2-8.1c0-88.4 71.6-160 160-160c59.3 0 111 32.2 138.7 80.2C409.9 102 428.3 96 448 96c53 0 96 43 96 96c0 12.2-2.3 23.8-6.4 34.6C596 238.4 640 290.1 640 352c0 70.7-57.3 128-128 128H144zm79-217c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l39-39V392c0 13.3 10.7 24 24 24s24-10.7 24-24V257.9l39 39c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-80-80c-9.4-9.4-24.6-9.4-33.9 0l-80 80z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="line-item-actions"
          >
            <button
              class="line-item-edit-button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-pencil "
                data-icon="pencil"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              class="line-item-remove-button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-trash "
                data-icon="trash"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="line-item-container"
    >
      <div
        class="line-item-container-left"
      >
        <div
          class="line-item-checkbox"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-square line-item-checkbox-unchecked"
            data-icon="square"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M384 80c8.8 0 16 7.2 16 16V416c0 8.8-7.2 16-16 16H64c-8.8 0-16-7.2-16-16V96c0-8.8 7.2-16 16-16H384zM64 32C28.7 32 0 60.7 0 96V416c0 35.3 28.7 64 64 64H384c35.3 0 64-28.7 64-64V96c0-35.3-28.7-64-64-64H64z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="line-item-name"
        >
          Shad B
        </div>
      </div>
      <div
        class="line-item-container-right"
      >
        <button
          class="line-item-upload-button"
        >
          <div
            class="line-item-upload-title"
          >
            Upload Letter
          </div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cloud-arrow-up line-item-upload-icon"
            data-icon="cloud-arrow-up"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M144 480C64.5 480 0 415.5 0 336c0-62.8 40.2-116.2 96.2-135.9c-.1-2.7-.2-5.4-.2-8.1c0-88.4 71.6-160 160-160c59.3 0 111 32.2 138.7 80.2C409.9 102 428.3 96 448 96c53 0 96 43 96 96c0 12.2-2.3 23.8-6.4 34.6C596 238.4 640 290.1 640 352c0 70.7-57.3 128-128 128H144zm79-217c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l39-39V392c0 13.3 10.7 24 24 24s24-10.7 24-24V257.9l39 39c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-80-80c-9.4-9.4-24.6-9.4-33.9 0l-80 80z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="line-item-actions"
        >
          <button
            class="line-item-edit-button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-pencil "
              data-icon="pencil"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="line-item-remove-button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-trash "
              data-icon="trash"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
